In [2]:
import pandas as pd
import sys

In [3]:
# below code is for getting the train data (70% of dataset) from github,

In [4]:
train_len = 382
columns = ['price', 'lotsize', 'bedrooms', 'bathrms']
traindata = pd.read_csv('https://raw.githubusercontent.com/iib2019014/ml_iib2019014_datasets/main/Housing%20Price%20data%20set.csv', nrows=train_len, usecols=columns)
# traindata.to_csv('traindata.csv')

testdata = pd.read_csv('https://raw.githubusercontent.com/iib2019014/ml_iib2019014_datasets/main/Housing%20Price%20data%20set.csv', skiprows=[i for i in range(1, train_len)], usecols=columns)
# testdata.to_csv('testdata.csv')

# %ls ""

In [5]:
# below code is for individually storing the features and price,

In [6]:
y, x1, x2, x3 = [], [], [], []

for i in range(train_len) :
  y.append(traindata['price'][i])
  x1.append(traindata['lotsize'][i])
  x2.append(traindata['bedrooms'][i])
  x3.append(traindata['bathrms'][i])

partialDerivate(W, wrt) function calculates the partial derivative of the error function with respective to the parameter w[wrt], where 'wrt' is the index in parameter list, W.

getError(W) function calculates the error with respective to the corresponding parameters list, W.

Gradient Descent Algorithm

In [7]:
# below cell has the functions to calculate the partial derivative and error,

In [8]:
def partialDerivate(w0, w1, w2, w3, wrt) :
  derivative_at = 0
  if wrt == 0 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]
    derivative_at /= m
    
  if wrt == 1 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x1[i]
    derivative_at /= m

  if wrt == 2 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x2[i]
    derivative_at /= m

  if wrt == 3 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x3[i]
    derivative_at /= m

  return derivative_at


def getError(w0, w1, w2, w3) :
  error = 0
  for i in range(m) :
    error += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) ** 2

  return error / (2 * m)

In the below cell, code for Batch-Gradient Descent Algo is writtem.
We expect the convergence will happend after 100 iterations.
Learning rate is taken to be 0.5

In [9]:
converge_limit = 45
learning_rate = 0.00000001
minerror = sys.maxsize
m = train_len
# w0, w1, w2, w3 = 0, 0, 0, 0
w0, w1, w2, w3 = 1, 1, 1, 1
b0, b1, b2, b3 = 0, 0, 0, 0


for i in range(converge_limit) :
  pd = partialDerivate(w0, w1, w2, w3, 0)
  t0 = w0 - learning_rate * pd
  # w0 = w0 - learning_rate * pd
  pd = partialDerivate(w0, w1, w2, w3, 1)
  t1 = w1 - learning_rate * pd
  # w1 = w1 - learning_rate * pd
  pd = partialDerivate(w0, w1, w2, w3, 2)
  t2 = w2 - learning_rate * pd
  # w2 = w2 - learning_rate * pd
  pd = partialDerivate(w0, w1, w2, w3, 2)
  t3 = w3 - learning_rate * pd
  # w3 = w3 - learning_rate * pd
  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getError(w0, w1, w2, w3)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

Minimum error is: 317496785.9328654
Corresponding parameters are: 
1.0039662991439684 12.513264188138088 1.0137825768806674 1.0137825768806674


# Analysis of error estimation using different learning rates

for learning rate = 0.00000001</br>
Minimum error is: 317527390.37990797</br>
Corresponding parameters are: 
0.006492306262885747, 12.514213155262858, 0.023658411497197745, 0.023658411497197745

for learning rate = 0.000000001</br>
Minimum error is: 326643185.95875156
Corresponding parameters are: 
0.002503928553911345 11.689990520090188 0.007933867079084573 0.007933867079084573

for learning rate = 0.0000001</br>
Minimum error is: 6272882965.908103</br>
Corresponding parameters are: 
0.006393635340314135, 33.58195581151832, 0.019611233246073296, 0.019611233246073296

for learning rate = 0.000001</br>
Minimum error is: 1402801609354.605
Corresponding parameters are: 
0.06393635340314135 335.8195581151832 0.19611233246073298 0.19611233246073298

In [10]:
# below cell calculates the error with batch gradient descent,

In [11]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2

print("error is " + str(error / (2 * len(testdata))))

error is 401132617.9874337


Stochastic Gradient Descent Algorithm

In [12]:
converge_limit = 100
learning_rate = 0.00001
minerror = sys.maxsize
m = train_len
w0, w1, w2, w3 = 0, 0, 0, 0
b0, b1, b2, b3 = 0, 0, 0, 0

for i in range(converge_limit) :
  k = i * m // converge_limit
  t0 = w0 - learning_rate * (w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) / m
  t1 = w1 - learning_rate * ((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x1[k]) / m
  t2 = w2 - learning_rate * ((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x2[k]) / m
  t3 = w3 - learning_rate * ((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x3[k]) / m

  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getError(w0, w1, w2, w3)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print(learning_rate)
print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

1e-05
Minimum error is: 317904561.91707104
Corresponding parameters are: 
0.007779951407274272 12.346619589370821 0.01652198301032323 0.00715471214079526


# Analysis of error estimation using different learning rates

for learning rate = 0.0001</br>
Minimum error is: 36327090871.993645</br>
Corresponding parameters are: 
0.01099476439790576, 64.31937172774869, 0.03298429319371728, 0.01099476439790576

for learning rate = 0.00001</br>
Minimum error is: 317904561.91707104</br>
Corresponding parameters are: 
0.007779951407274272, 12.346619589370821, 0.01652198301032323, 0.00715471214079526

for learning rate = 0.000001</br>
Minimum error is: 317527896.40058625</br>
Corresponding parameters are: 
0.003399376904692982, 12.516202404263344, 0.00892773354359082, 0.0039020307829233025

for learning rate = 0.0000001</br>
Minimum error is: 1092576692.2315826</br>
Corresponding parameters are: 
0.0010938203180438553, 4.913967141761904, 0.003047065368094102, 0.001325219280389634

In [13]:
# below cell calculates the error with stochastic gradient descent,

In [14]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2

print("error is " + str(error / (2 * len(testdata))))

error is 397054274.24884486


Mini-Batch Gradient Descent Algorithm

In [44]:
converge_limit = 100
learning_rate = 0.00001
minerror = sys.maxsize
m = train_len
w0, w1, w2, w3 = 0, 0, 0, 0
b0, b1, b2, b3 = 0, 0, 0, 0

for i in range(converge_limit) :
  j = i * m // converge_limit
  sum0, sum1, sum2, sum3 = 0, 0, 0, 0
  for k in range(10) :
    if j + k >= train_len :
      break
    sum0 += w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]
    sum1 += (w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x1[k]
    sum2 += (w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x2[k]
    sum3 += (w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x3[k]
  t0 = w0 - learning_rate * sum0 / m
  t1 = w1 - learning_rate * sum2 / m
  t2 = w2 - learning_rate * sum2 / m
  t3 = w3 - learning_rate * sum3 / m

  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getError(w0, w1, w2, w3)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print(learning_rate)
print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

1e-05
Minimum error is: 1443701447.1143599
Corresponding parameters are: 
1.1955455496463587 3.350081055970454 3.350081055970454 1.4351759732254343


# Analysis of error estimation using different learning rates

for learning rate = 0.01</br>
Minimum error is: 4806592020.676374</br>
Corresponding parameters are: 
10.99476439790576, 30.785340314136125, 30.785340314136125, 13.193717277486911

for learning rate = 0.001</br>
Minimum error is: 317251564.95131403</br>
Corresponding parameters are: 
4.012063537369557, 12.487660475302741, 12.487660475302741, 5.3778116726440155

for learning rate = 0.0001</br>
Minimum error is: 319943740.2773181</br>
Corresponding parameters are: 
4.260235075117864, 12.058207614163068, 12.058207614163068, 5.170485407624245

for learning rate = 1e-05</br>
Minimum error is: 1443701447.1143599</br>
Corresponding parameters are: 
1.1955455496463587 3.350081055970454 3.350081055970454 1.4351759732254343

In [16]:
# below cell calculates the error with mini-batch gradient descent,

In [17]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2

print("error is " + str(error / (2 * len(testdata))))

error is 1882824970.8933487


# With Regularization

In [18]:
# below cell has the functions for calculating the partial derivates and error, but here along with regularization term,

In [19]:
def partialDerivateReg(w0, w1, w2, w3, wrt, hyplambda) :
  derivative_at = 0
  if wrt == 0 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i] + (hyplambda * w0)
    derivative_at /= m
    
  if wrt == 1 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x1[i] + (hyplambda * w1)
    derivative_at /= m

  if wrt == 2 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x2[i] + (hyplambda * w2)
    derivative_at /= m

  if wrt == 3 :
    derivative_at = 0
    for i in range(m) :
      derivative_at += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) * x3[i] + (hyplambda * w3)
    derivative_at /= m

  return derivative_at


def getErrorReg(w0, w1, w2, w3, hyplambda) :
  error = 0
  for i in range(m) :
    error += (w0 + w1 * x1[i] + w2 * x2[i] + w3 * x3[i] - y[i]) ** 2

  error +=  hyplambda * (w0 ** 2 + w1 ** 2 + w2 ** 2 + w3 ** 2)

  return error / (2 * m)

Batch-Gradient Descent Algorithm

In [45]:
converge_limit = 100
learning_rate = 0.0000001
minerror = sys.maxsize
m = train_len
w0, w1, w2, w3 = 0, 0, 0, 0
b0, b1, b2, b3 = 0, 0, 0, 0
hyplambda = 0.0005


for i in range(converge_limit) :
  pd = partialDerivateReg(w0, w1, w2, w3, 0, hyplambda)
  t0 = w0 - learning_rate * pd
  pd = partialDerivateReg(w0, w1, w2, w3, 1, hyplambda)
  t1 = w1 - learning_rate * pd
  pd = partialDerivateReg(w0, w1, w2, w3, 2, hyplambda)
  t2 = w2 - learning_rate * pd
  pd = partialDerivateReg(w0, w1, w2, w3, 2, hyplambda)
  t3 = w3 - learning_rate * pd
  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getErrorReg(w0, w1, w2, w3, hyplambda)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

Minimum error is: 6272882965.908841
Corresponding parameters are: 
0.006393635340314135 33.58195581151832 0.019611233246073296 0.019611233246073296


lambda = 0.0005</br>
Minimum error is: 317527390.38001066</br>
Corresponding parameters are: 
0.006492306261787789, 12.514213155029692, 0.023658411492870034, 0.023658411492870034

lambda = 0.5</br>
Minimum error is: 317527390.4825164</br>
Corresponding parameters are: 
0.006492305164926526, 12.514212922096618, 0.023658407169479923, 0.023658407169479923

lambda = 1.5</br>
Minimum error is: 317527390.6877382</br>
Corresponding parameters are: 
0.006492302969009118, 12.514212455764167, 0.0236583985140484, 0.0236583985140484

In [21]:
# below cell calculates the error with batch gradient descent using regularization,

In [22]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2

print("error is " + str(error / (2 * len(testdata))))

error is 401180244.65862715


Stochastic Gradient Descent Algorithm

In [51]:
converge_limit = 100
learning_rate = 0.000001
minerror = sys.maxsize
m = train_len
w0, w1, w2, w3 = 0, 0, 0, 0
b0, b1, b2, b3 = 0, 0, 0, 0
hyplambda = 0.005

for i in range(converge_limit) :
  k = i * m // converge_limit
  t0 = w0 - learning_rate * 1/m * ((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) + (hyplambda * w0))
  t1 = w1 - learning_rate * 1/m * (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x1[k]) + (hyplambda * w1))
  t2 = w2 - learning_rate * 1/m * (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x2[k]) + (hyplambda * w2))
  t3 = w3 - learning_rate * 1/m * (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x3[k]) + (hyplambda * w3))

  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getError(w0, w1, w2, w3)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print(learning_rate)
print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

1e-06
Minimum error is: 317527896.4004116
Corresponding parameters are: 
0.003399376903808568 12.516202400974228 0.008927733541568457 0.0039020307820025117


for learning rate = 1e-05</br>
Minimum error is: 317904561.92808753</br>
Corresponding parameters are: 
0.007779951395284366 12.346619586921399 0.016521982991642467 0.007154712133347393

for learning rate = 1e-06</br>
Minimum error is: 317527896.4004116</br>
Corresponding parameters are: 
0.003399376903808568 12.516202400974228 0.008927733541568457 0.0039020307820025117

for learning rate = 1e-07</br>
Minimum error is: 1092576692.2807183</br>
Corresponding parameters are: 
0.0010938203179799088 4.91396714152099 0.0030470653679212675 0.001325219280312025

for learning rate = 1e-08</br>
Minimum error is: 2203959170.177111</br>
Corresponding parameters are: 
0.00013793070332628532 0.6569651734812031 0.0003882756382303034 0.00017021293909980282

In [24]:
# below cell calculates the error with stochastic gradient descent using regularization,

In [25]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2

print("error is " + str(error / (2 * len(testdata))))

error is 2980821982.2258143


Mini-Batch Gradient Descent Algorithm

In [26]:
converge_limit = 100
learning_rate = 0.001
minerror = sys.maxsize
m = train_len
w0, w1, w2, w3 = 0, 0, 0, 0
b0, b1, b2, b3 = 0, 0, 0, 0

for i in range(converge_limit) :
  j = i * m // converge_limit
  sum0, sum1, sum2, sum3 = 0, 0, 0, 0
  for k in range(10) :
    if j + k >= train_len :
      break
    sum0 += ((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) + (hyplambda * w0))
    sum1 += (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x1[k]) + (hyplambda * w1))
    sum2 += (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x2[k]) + (hyplambda * w2))
    sum3 += (((w0 + w1 * x1[k] + w2 * x2[k] + w3 * x3[k] - y[i]) * x3[k]) + (hyplambda * w3))
  t0 = w0 - learning_rate * sum0 / m
  t1 = w1 - learning_rate * sum2 / m
  t2 = w2 - learning_rate * sum2 / m
  t3 = w3 - learning_rate * sum3 / m

  w0, w1, w2, w3 = t0, t1, t2, t3
  error = getError(w0, w1, w2, w3)
  if error < minerror :
    minerror = error
    b0, b1, b2, b3 = w0, w1, w2, w3

print(learning_rate)
print("Minimum error is: " + str(minerror))
print("Corresponding parameters are: ")
print(b0, b1, b2, b3)

0.001
Minimum error is: 317251566.9762469
Corresponding parameters are: 
4.012064149425027 12.487656441938277 12.487656441938277 5.377809825089577


0.001
Minimum error is: 317251566.9762469
Corresponding parameters are: 
4.012064149425027 12.487656441938277 12.487656441938277 5.377809825089577

0.0001
Minimum error is: 319943781.2928109
Corresponding parameters are: 
4.260233892792104 12.058204208300243 12.058204208300243 5.1704839459946434

1e-05
Minimum error is: 1443701495.4302328
Corresponding parameters are: 
1.195545479608556 3.3500808596056912 3.3500808596056912 1.4351758890314188

1e-06
Minimum error is: 2290456823.9002075
Corresponding parameters are: 
0.13853420698045682 0.3878714655534293 0.3878714655534293 0.1661181362496273

In [27]:
# below cell calculates the error with mini-batch gradient descent using regularization,

In [28]:
error = 0
for i in range(len(testdata)) :
    error += (testdata['price'][i] - (b0 + b1 * testdata['lotsize'][i] + b2 * testdata['bedrooms'][i] + b3 * testdata['bathrms'][i])) ** 2
    
print("error is " + str(error / (2 * len(testdata))))

error is 400253671.9536608


# QUESTION 6d

In [29]:
# beow cell has imports of numpy and math to solve the matrix operations and exponential funciton,

In [30]:
import numpy as np
import math

In [31]:
# below function calculates the weight for each sample point from the query point,
# test1, test2, test3 are the features of the query point,
# train1, train2, train3 are the features of the sample point,

In [32]:
def calcWeight(test1, test2, test3, train1, train2, train3, tau) :
    distanceSquare = (test1 - train1) ** 2 + (test2 - train2) ** 2 + (test3 - train3) ** 2
    e = math.exp(-(distanceSquare/(2 * tau ** 2)))
    return e

In [33]:
# below cell calculates the optimal parameters using the locally weighted regression technique,
# different tau values are selected,

In [41]:
X = [[1 for i in range(train_len)], x1, x2, x3]
weights = [[0 for i in range(train_len)] for j in range(train_len)]
tau = 310
error = 0

for i in range(len(testdata)) :
    for j in range(train_len) :
        weights[j][j] = calcWeight(testdata['lotsize'][i], testdata['bedrooms'][i], testdata['bathrms'][i], traindata['lotsize'][j], traindata['bedrooms'][j], traindata['bathrms'][j], tau)
    w = np.dot(np.linalg.inv(np.dot(np.dot(X, weights), np.transpose(X))), (np.dot(np.dot(X, weights), y)))
    error += (testdata['price'][i] - (w[0] + w[1] * testdata['lotsize'][i] + w[2] * testdata['bedrooms'][i] + w[3] * testdata['bathrms'][i])) ** 2
error

print(w)
# print(error)
print(error / (2 * len(testdata)))

[-6.34104742e+04  1.34586300e+01  9.22778922e+03  2.33104689e+04]
174319420.69389585


tau = 290</br>
[-6.14719383e+04  1.30014834e+01  9.65695268e+03  2.30397487e+04]</br>
309769532.58574605

tau = 295</br>
[-6.20593384e+04  1.31343009e+01  9.54468663e+03  2.31119248e+04]</br>
178153536.31572813</br>

tau = 300</br>
[-6.25767508e+04  1.32545159e+01  9.43575495e+03  2.31810297e+04]</br>
162797522.4925398</br>

tau = 305</br>
[-6.30263680e+04  1.33624965e+01  9.33013372e+03  2.32471749e+04]</br>
276294199.9975849</br>

tau = 310</br>
[-6.34104742e+04  1.34586300e+01  9.22778922e+03  2.33104689e+04]</br>
174319420.69389585</br>